docker/daemon/client.go

```diff
@@ -0,0 +1,69 @@
+package daemon
+
+import (
+	"github.com/containers/image/types"
+	dockerclient "github.com/docker/docker/client"
+	"github.com/docker/go-connections/tlsconfig"
+	"net/http"
+	"path/filepath"
+)
+
+const (
+	// The default API version to be used in case none is explicitly specified
+	defaultAPIVersion = "1.22"
+)
+
+// NewDockerClient initializes a new API client based on the passed SystemContext.
+func newDockerClient(ctx *types.SystemContext) (*dockerclient.Client, error) {
+	if ctx == nil {
+		return dockerclient.NewClient(dockerclient.DefaultDockerHost, defaultAPIVersion, nil, nil)
+	}
+
+	httpClient, err := tlsConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	host := ctx.DockerDaemonHost
+	if host == "" {
+		host = dockerclient.DefaultDockerHost
+	}
+
+	version := ctx.DockerDaemonAPIVersion
+	if version == "" {
+		version = defaultAPIVersion
+	}
+
+	cli, err := dockerclient.NewClient(host, version, httpClient, nil)
+	if err != nil {
+		return cli, err
+	}
+
+	return cli, nil
+}
+
+func tlsConfig(ctx *types.SystemContext) (*http.Client, error) {
+	if ctx == nil || ctx.DockerDaemonCertPath == "" {
+		return nil, nil
+	}
+
+	options := tlsconfig.Options{
+		CAFile:             filepath.Join(ctx.DockerDaemonCertPath, "ca.pem"),
+		CertFile:           filepath.Join(ctx.DockerDaemonCertPath, "cert.pem"),
+		KeyFile:            filepath.Join(ctx.DockerDaemonCertPath, "key.pem"),
+		InsecureSkipVerify: ctx.DockerDaemonInsecureSkipTLSVerify,
+	}
+	tlsc, err := tlsconfig.Client(options)
+	if err != nil {
+		return nil, err
+	}
+
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsc,
+		},
+		CheckRedirect: dockerclient.CheckRedirect,
```

Comment:
It seems strange that this `CheckRedirect` value is set only if `ctx.DockerDaemonCertPath != ""`.

Reply:
If it is "", then we are back to the case where we pass no http.client into client.NewClient and CheckRedirect occurs there:

```go
	if client != nil {
		if _, ok := client.Transport.(http.RoundTripper); !ok {
			return nil, fmt.Errorf("unable to verify TLS configuration, invalid transport %v", client.Transport)
		}
	} else {
		transport := new(http.Transport)
		sockets.ConfigureTransport(transport, proto, addr)
		client = &http.Client{
			Transport:     transport,
			CheckRedirect: CheckRedirect,
		}
	}
```

Reply:
Ah, true.  But just looking at the above, the `ctx.DockaerDaemonCertPath` is missing the `sockets.ConfigureTransport` call (i.e. `*FromEnvironment` support).

Having two paths with which can (i.e. long-term will) diverge like this just seems painful.

If all that we _want_ to edit is the TLS configuration, the cleanest way to do that is if the only difference between the settings we supply is the TLS configuration.

Reply:
I think this will be implicitly fixed by the change I am making for https://github.com/containers/image/pull/328#discussion_r148012858

Reply:
That change removed the difference caused by `ctx.DockerDaemonCertPath`, but the behavior still depends on `ctx != nil`.

Really, figuring that the various paths are equivalent is too hard to write, let alone to maintain.  _Please_ just implement exactly one path; the cost of a few `ctx != nil` checks in `if ctx != nil && ctx.DockerDaemon… {}` isn’t that much.

Reply:
> That change removed the difference caused by ctx.DockerDaemonCertPath, but the behavior still depends on ctx != nil.

right, if I don't get a context, nil is returned. This is different from returning tlsconfig.Options{}.

> if ctx != nil && ctx.DockerDaemon… {} isn’t that much.

This will guard against a panic and potentially accessing _DockerDaemon_ in case ctx is nil, but at some stage I still would have to return nil or a ctx.DockerDaemon instance (provided I want to keep the semantics of the function)

Reply:
> This is different from returning tlsconfig.Options{}.

Yes, but it is defined to be the same.  `docker-daemon:` behavior should not change just because the caller needed to define a completely `docker-daemon:`-unrelated setting like `SignaturePolicyPath`.

> but at some stage I still would have to return nil or a ctx.DockerDaemon instance (provided I want to keep the semantics of the function)

I don’t understand what you mean by this. `ctx == nil` and `ctx.DockerDaemon… /* is unset */` should always behave precisely equivalently.

```go
// NewDockerClient initializes a new API client based on the passed SystemContext.
func newDockerClient(ctx *types.SystemContext) (*dockerclient.Client, error) {
	httpClient, err := tlsConfig(ctx)
	if err != nil {
		return nil, err
	}

	host := dockerclient.DefaultDockerHost
	if ctx != nil & ctx.DockerDaemonHost != "" {
		host = ctx.DockerDaemonHost
	}

	return dockerclient.NewClient(host, defaultAPIVersion, httpClient, nil)
}

func tlsConfig(ctx *types.SystemContext) (*http.Client, error) {
	options := tlsconfig.Options{}
	if ctx != nil && ctx.DockerDaemonInsecureSkipTLSVerify {
		options.InsecureSkipVerify = true
	}

	if ctx != nil && ctx.DockerDaemonCertPath != "" {
		options.CAFile = filepath.Join(ctx.DockerDaemonCertPath, "ca.pem")
		options.CertFile = filepath.Join(ctx.DockerDaemonCertPath, "cert.pem")
		options.KeyFile = filepath.Join(ctx.DockerDaemonCertPath, "key.pem")
	}

	tlsc, err := tlsconfig.Client(options)
	if err != nil {
		return nil, err
	}

	return &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: tlsc,
		},
		CheckRedirect: dockerclient.CheckRedirect,
	}, nil
}
```
?

Reply:
Right, this is nicer. However, this also changes a bit of the semantics. Now you always create an http.Client which is passed to dockerclient.NewClient. The original code was passing nil. My intent was to keep this in place. dockerclient.NewClient does actually do a check whether httpClient is nil and if so get the TLS config from the default transport. I wanted to avoid any potential change in behaviour. I am good with this as well, but just want to point out the difference.

Reply:
> I wanted to avoid any potential change in behaviour.

That needs to be balanced with 1) `&types.SystemContext{}` not changing behavior, and 2) non-empty `SystemContext{DockerDaemonSomething}` not changing unrelated aspects of the configuration.

I agree it would be nice to be able to get a `http.Client` exactly equivalent to passing `nil` out of the `…/docker/client` package, but that’s not available AFAICT.

Reply:
Sure. It's always a trade-off. So we are good now?